Basic utilities to support remote autotuning

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #153201
* #153486

Summary: Adds some new utilities in autotune_remote.py that will be used to construct remote autotuning requests (tensor metadata, extern kernel details, triton kernel details). For Triton, I reuse the existing    TritonBenchmarkRequest / TritonTemplateCaller that we use for subprocess autotuning. For extern kernels, I provide the name of the kernel and expect to look it up on the other side of the wire. The new utilities also contain the methods to construct args from the serialized request and call the existing ("in process") benchmark helpers in AlgorithmSelectorCache.

Test Plan: New test case that runs all the test_select_algorithm tests, but routes the benchmark request through the remote utilties to make sure the remote request a) is serializable, and b) the choices in the deserialized object are benchmarkable.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

masnesral committed:
Basic utilities to support remote autotuning

Summary: Adds some new utilities in autotune_remote.py that will be used to construct remote autotuning requests (tensor metadata, extern kernel details, triton kernel details). For Triton, I reuse the existing    TritonBenchmarkRequest / TritonTemplateCaller that we use for subprocess autotuning. For extern kernels, I provide the name of the kernel and expect to look it up on the other side of the wire. The new utilities also contain the methods to construct args from the serialized request and call the existing ("in process") benchmark helpers in AlgorithmSelectorCache.

Test Plan: New test case that runs all the test_select_algorithm tests, but routes the benchmark request through the remote utilties to make sure the remote request a) is serializable, and b) the choices in the deserialized object are benchmarkable.

[ghstack-poisoned]

eellison commented on torch/_inductor/autotune_remote.py:
```diff
@@ -0,0 +1,222 @@
+    view_size: ShapeType
+    view_stride: StrideType
+    view_offset: int
+    allocation_size: Sequence[sympy.Expr]
+    name: str
+    value: Optional[torch.Tensor] = None
```
How does this work with sympy expressions ? the remote server wont have the shape env. We'd want to at least pass hints in I think. 

Also, we probably want the invariant that the benchmarking wont add any guards.

aorenste replied:
I don't know enough about autotuning - but can autotuning use sympy expressions? That's really only valid with FakeTensors, right? I would expect that autotuning would have to happen on real tensors since you're looking at real performance? Or am I missing something?

masnesral replied:
synced offline. We're calling get_hints() to get the allocation_size, so this should be the same type as the size field.

eellison commented on test/inductor/test_select_algorithm.py:
```diff
@@ -369,6 +371,40 @@ def test_TritonTemplateCaller_str(self):
+    return benchmark
+
+
+class TestSelectAlgorithmRemote(TestSelectAlgorithm):
```
nice ! it might be worth doing this on a subset of the test_max_autotune tests as well

eellison reviewed (commented):
one question about symbolic shapes

aorenste commented on torch/_inductor/autotune_remote.py:
```diff
@@ -0,0 +1,222 @@
+
+
+@dataclasses.dataclass
+class TensorMeta:
```
I wonder if we could/should reuse torch._subclasses.meta_utils.MetaTensorDesc

masnesral committed:
Update on "Basic utilities to support remote autotuning"


Summary: Adds some new utilities in autotune_remote.py that will be used to construct remote autotuning requests (tensor metadata, extern kernel details, triton kernel details). For Triton, I reuse the existing    TritonBenchmarkRequest / TritonTemplateCaller that we use for subprocess autotuning. For extern kernels, I provide the name of the kernel and expect to look it up on the other side of the wire. The new utilities also contain the methods to construct args from the serialized request and call the existing ("in process") benchmark helpers in AlgorithmSelectorCache.

Test Plan: New test case that runs all the test_select_algorithm tests, but routes the benchmark request through the remote utilties to make sure the remote request a) is serializable, and b) the choices in the deserialized object are benchmarkable.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
